### ISSUE TITLE
Resolve connectivity issues with default Multus networks on the GCP test setup

### ISSUE THREAD

**Issue by stoyanr** [gardener/gardener-extension-provider-kubevirt#31]
When using a GCP cluster as a provider cluster, as described in [Development Setup](docs/dev-setup.md), we are so far not able to perform a successful test with a default Multus network, even though the network seems to be configured correctly and works in other environments.

The default Multus network uses the `bridge` CNI plugin with the following configuration:

```json
{
  "cniVersion": "0.3.1",
  "type": "bridge",
  "master": "ens4",
  "isGateway": true,
  "isDefaultGateway": true,
  "ipMasq": true,
  "ipam": {
    "type": "host-local",
    "subnet": "192.168.0.0/16"
  }
}
```

**Note:** The bridge plugin is not available by default on the GCP host and has to be installed manually:

```sh
wget https://github.com/containernetworking/plugins/releases/download/v0.8.6/cni-plugins-linux-amd64-v0.8.6.tgz
tar -xvf cni-plugins-linux-amd64-v0.8.6.tgz
cp bridge /opt/cni/bin/
```

When the VM starts, we see the network interface having the correct IP address, and we also see the `cni0` bridge interface on the GCP host. However, the `cloud-init` fails due to the VM being unable to connect to the Internet with errors similar to:

```
[   93.675859] cloud-init[1114]: W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/bionic-security/InRelease  Cannot initiate the connection to security.ubuntu.com:80 (2001:67c:1562::15). - connect (101: Network is unreachable) ..., connection timed out
```

We checked that if we create a pod manually connected to the same network (via the `v1.multus-cni.io/default-network` annotation), the pod is also unable to reach the Internet. Also, the pod and the VM are not able to ping each other.

We checked also multiple other things but were so far not able to find the reason for this failure. Some of the things we tried:

* Checking network interfaces and IP addresses on the GCP host with `ip a`.
* Checking routes with `route` on the GCP host.
* Checking routes with `route` on the Kubevirt VM.
* Checking IP forwarding with `sysctl net.ipv4.ip_forward` on the GCP host, response is `net.ipv4.ip_forward = 1`.
* Running `arp -n` on the GCP host, able to see all IP addresses correctly attached to their network interfaces.
* Checking iptables rules on the GCP host with `iptables -L | grep -i -e "REJECT\|DROP"` and `iptables -L -t nat` (look for 192.168.*), no suspicious iptables rules found.
* Checking GCP firewall rules - no Egress rules are present.
* Check Kubernetes network policies - none present in the `default` namespace of the provider cluster.
* Checking if the promiscuous mode is enabled on the network interface with `brctl show cni0` and then `ip -d link show veth78574d57`, promiscuous mode is enabled as it should be.

We should find the reason for this failure and fix it, as it would be very beneficial to be able to test default Multus networks in such test setups. This could be based on `bridge`, `macvlan`, or any other CNI plugin, as far as it works correctly.

/area networking
/kind task
/priority normal
/platform kubevirt

**Comment by mfranczy** [comment]
It seems that nat rules for the bridge were not set. I am going to confirm that, but first I am trying to create the same gcp environemnt, I have a few problems so far with gardener...

**Comment by mfranczy** [comment]
I've created k8s cluster on GCP and I have internet connection from vms (I deployed kubevirt vms on GCP vm), bellow my manifests

NAD:
```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: net-conf
  namespace: default
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "bridge",
      "isGateway": true,
      "isDefaultGateway": true,
      "ipMasq": true,
      "ipam": {
        "type": "host-local",
        "subnet": "192.168.0.0/16"
        }
    }'
```

**`isDefaultGateway` must be true**

from my vm (created by kubevirt)
```
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    100    0        0 enp1s0
192.168.0.0     0.0.0.0         255.255.0.0     U     0      0        0 enp1s0
```

NAT is appropriate on the GCP VM
```
root@mfranczy-test-control-plane-2:~# iptables -L -tnat | grep net-conf
CNI-7c96404ba3c392cdd43ad191  all  --  192.168.0.2          anywhere             /* name: "net-conf" id: "e11ad4d16074b9cc21a57794ce9208ab544857a26c52f75046a99d47d5468706" */
CNI-6b6ad2394f9a7b4d75fa1b22  all  --  192.168.0.3          anywhere             /* name: "net-conf" id: "a63eba176306aec4171a01129a21ee04f84fbc3c2e0ffb0d0807bb03988a779c" */
CNI-8a12a2d356eba227afc441ca  all  --  192.168.0.4          anywhere             /* name: "net-conf" id: "da6460ed5463e7f79e94f7b176c56d29460f914f865719b088b7c32485677f3a" */
CNI-819d5d942dffb1489b33da11  all  --  192.168.0.5          anywhere             /* name: "net-conf" id: "4edc968e37a70919e027867c9632fcf9c5c4c8b78be22a4ef6c047314c9a96f4" */
CNI-22d3fd4e2d7df2540cc1bf9e  all  --  192.168.0.6          anywhere             /* name: "net-conf" id: "5532bbca69dc79a22f525f1dca44d95d7bb0598d463e4395edb8eab92fd71e5b" */
ACCEPT     all  --  anywhere             192.168.0.0/16       /* name: "net-conf" id: "5532bbca69dc79a22f525f1dca44d95d7bb0598d463e4395edb8eab92fd71e5b" */
MASQUERADE  all  --  anywhere            !base-address.mcast.net/4  /* name: "net-conf" id: "5532bbca69dc79a22f525f1dca44d95d7bb0598d463e4395edb8eab92fd71e5b" */
ACCEPT     all  --  anywhere             192.168.0.0/16       /* name: "net-conf" id: "a63eba176306aec4171a01129a21ee04f84fbc3c2e0ffb0d0807bb03988a779c" */
MASQUERADE  all  --  anywhere            !base-address.mcast.net/4  /* name: "net-conf" id: "a63eba176306aec4171a01129a21ee04f84fbc3c2e0ffb0d0807bb03988a779c" */
ACCEPT     all  --  anywhere             192.168.0.0/16       /* name: "net-conf" id: "e11ad4d16074b9cc21a57794ce9208ab544857a26c52f75046a99d47d5468706" */
MASQUERADE  all  --  anywhere            !base-address.mcast.net/4  /* name: "net-conf" id: "e11ad4d16074b9cc21a57794ce9208ab544857a26c52f75046a99d47d5468706" */
ACCEPT     all  --  anywhere             192.168.0.0/16       /* name: "net-conf" id: "4edc968e37a70919e027867c9632fcf9c5c4c8b78be22a4ef6c047314c9a96f4" */
MASQUERADE  all  --  anywhere            !base-address.mcast.net/4  /* name: "net-conf" id: "4edc968e37a70919e027867c9632fcf9c5c4c8b78be22a4ef6c047314c9a96f4" */
ACCEPT     all  --  anywhere             192.168.0.0/16       /* name: "net-conf" id: "da6460ed5463e7f79e94f7b176c56d29460f914f865719b088b7c32485677f3a" */
MASQUERADE  all  --  anywhere            !base-address.mcast.net/4  /* name: "net-conf" id: "da6460ed5463e7f79e94f7b176c56d29460f914f865719b088b7c32485677f3a" */
```

routing as well (still GCP vm)

```
root@mfranczy-test-control-plane-2:~# route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    100    0        0 ens4
_gateway        0.0.0.0         255.255.255.255 UH    100    0        0 ens4
...
192.168.0.0     0.0.0.0         255.255.0.0     U     0      0        0 cni0
```

curl from kubevirt VM
```
ubuntu@vm-worker:~$ curl http://google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
ubuntu@vm-worker:~$
```

GCP itself is not a problem, we have to revisit your GCP project settings or network policies from gardener.
There is no other option.

**Comment by mfranczy** [comment]
@stoyanr, I am going to wait with further debugging till you're back from vacation as I cannot recreate the issue.

**Comment by mfranczy** [comment]
probably it's because of calico setup and the thing that the pod has an IP address outside of the cluster CIDR

**Comment by stoyanr** [comment]
We continued debugging and identified that the reason for the original issue was missing iptables rules to allow traffic to/from the container. We are not really sure why it's not reproducible on some test setups, we suspect it has something to do with the cluster CNI plugin used (calico vs. flannel). Eventually, we were able to overcome these issues by using the [firewall](https://github.com/containernetworking/plugins) CNI plugin in addition to `bridge`, with the following configuration:

```yaml
{
	"cniVersion": "0.4.0",
	"name": "bridge-firewall",
	"plugins": [
		{
			"type": "bridge",
			"isGateway": true,
			"isDefaultGateway": true,
			"ipMasq": true,
			"ipam": {
				"type": "host-local",
				"subnet": "10.100.0.0/16"
			}
		},
		{
			"type": "firewall"
		}
	]
}
```

With the above configuration, the issue is resolved and we are able to fully reconcile test clusters on GCP test setups with both calico and cilium as cluster CNI plugin.
